Add parameter_space module

schmoelder opened the pull request:
See also #272 

## Todo
- [x] Bounds
- [x] Parameter normalization
- [x] Linear constraints
- [x] Linear equality constraints
- [~] Parameter dependencies
- [ ] Parameter mapping (i.e. how to map parameters to actual model parameters, providing different setter methods etc.)
- [ ] Mapping with indices, see also: [forum post](https://forum.cadet-web.de/t/specifying-indices-of-multidimensional-parameters-for-events-and-optimizationvariables/755) 
    - [ ] multidimensional arrays
    - [ ] `np.s_`
    - [ ] polynomial parameters
- [ ] Modularize parameter_space.py 
- [ ] Modularize tests for parameter_space.py
- [ ] Add NormalizedParameterSpace class
- [ ] Sampling methods (e.g. with hopsy)


## Open questions
- [x ] Where does normalization "live"? Parameter vs ParameterSpace? -> Parameter (e.g. choice parameters cannot be normalized anyway)
- [x] Where do dependencies "live"? Parameter vs ParameterSpace? Currently, they live in the Parameter. But we could consider moving them to the Parameter space to keep the Parameter class simple.
- [ ] Do we want to allow a more "dynamic" configuration? E.g. add_dependencies after instantiation?

schmoelder reviewed (commented):
Hey, good work. Note, I had no problem running the tests, neither in Spyder, nor in my terminal. So not sure why it didn't work for you. If you want, we can have another look together.

talasunna commented on tests/parameter_space/test_parameters.py:
```diff
@@ -0,0 +1,105 @@
+    p = RangedParameter(name="x", lb=0, ub=10, parameter_type=float)
+    pspace._parameters.append(p)
+    pspace.add_linear_constraint(parameters=p, a=2.0, b=10.0)
+    assert len(pspace.linear_constraints) == 1
```
I added the add_linear_constraint method in the parameters.py module and fixed the rest of the tests, they all pass now so I think the issue is hopefully solved 🤓

schmoelder replied:
:rocket: 

I can still see errors, however. 

```
FAILED tests/parameter_space/test_parameters.py::test_linear_constraint_invalid_length - ValueError: Length of lhs must match number of parameters.
FAILED tests/parameter_space/test_parameters.py::test_parameter_space_add_linear_constraint - TypeError: ParameterSpace.add_linear_constraint() got an unexpected keyword argument 'a'
```

schmoelder commented:
Here are some notes from today's call:

```
class ParameterMapper():
    def __init__(self, parameter, eval_obj):
        self.eval_obj = eval_obj

    def set_value(self, value):
        raise NotImplementedError()


class ParameterAttributeSetter():
    def __init__(self, parameter, eval_obj, attribute):
        super().__init__(parameter, eval_obj)
        self.attribute = attribute

    def set_value(self, value):
        setattr(self.eval_obj, self.attribute, value)


class ParameterNestedDictSetter(ParameterMapper):
    def __init__(self, parameter, eval_obj, parameter_path):
        super().__init__(parameter, eval_obj)
        self.parameter_path = parameter_path

    def set_value(self, value):
        parameters = generate_nested_dict(self.parameter_path, value)
        self.eval_obj.parameters = parameters


# %% Old way of configuring optimization variables in evaluation objects
process_1 = setup_process("slow")
process_2 = setup_process("fast")

optimization_problem.add_evaluation_object(process_1)
optimization_problem.add_evaluation_object(process_2)

optimization_problem.add_variable(
    name='k_a'
    parameter_path="flow_sheet.column_1.binding_model.k_a",
    lb=1e-6, ub=1e-3,
    transform="log"
)


# %% New way

process_1 = setup_process("slow")
process_2 = setup_process("fast")

variable_1 = Parameter("k_a", lb=1e-6, ub=1e-3)

# Option 1
parameter_mapper = ParameterNestedDictSetter(
    variable_1,
    [process_1, process_2],
    "flow_sheet.column_1.binding_model.k_a"
)

# Option 2
parameter_mapper = ParameterAttributeSetter(
    variable_1,
    [process_1.flow_sheet.column, process_2.flow_sheet.column],
    "k_a" # Should we use the name if it's not provided?
)
```

Open questions:
- How does this connect to the ParameterSpace?
- How to integrate this into the OptimizationProblem?
- How to integrate "indices"?

schmoelder commented on CADETProcess/parameter_space/parameters.py:
```diff
@@ -0,0 +1,381 @@
+        return self._parameters
+
+    @property
+    def dependent_parameters(self) -> list[ParameterBase]:
```
Also add independent parameters property.

schmoelder commented on CADETProcess/parameter_space/parameters.py:
```diff
@@ -0,0 +1,381 @@
+        """
+        Validate the value.
+
+        Iterates over parameter mappers and sets the value.
```
Add a `TODO` here.

Commit:
Add normalize module and normalize tests

Commit:
Add parameters module and parameters test

schmoelder committed:
fixup! Add parameters module and parameters test

schmoelder committed:
[WIP] Add method to get dependent values

schmoelder committed:
[WIP] Add mapping

schmoelder committed:
Add todo

Commit:
refactor: introduce hybrid dot-path setter and remove descend helpers

* Added `_traverse_mixed_path` + `ParameterDotPathSetter` that walk a
  dot-separated path across **dict keys and object attributes** in one
go.
  - no automatic branch creation; missing hops raise
KeyError/AttributeError
  - uses `collections.abc.Mapping` to accept any dict-like container
* Dropped obsolete `descend()` and `set_nested_value()` utilities; logic
is
  now local to the class for better readability and maintenance.
* Tightened type hints (`Sequence[str]` for path tokens) and enabled
  `@dataclass(slots=True)` for memory-efficient instances.

Commit:
added tests for new mapping functionality and edited mapping again

schmoelder commented on tests/parameter_space/test_parameters.py:
```diff
@@ -0,0 +1,361 @@
+
+
+@pytest.fixture
+def random_dict():
```
It's not really random, right? Try to be more specific, e.g. call it "nested_dict".

schmoelder commented on tests/parameter_space/test_parameters.py:
```diff
@@ -0,0 +1,361 @@
+
+@pytest.fixture
+def random_dict():
+    """Dictionary mimicking a flowsheet‐like structure."""
```
[Nitpicking] While we will use this for objects such as flow-sheet, your method works with any kind of deeply nested structure. So you could also chose more generic parameter names. Sometimes this helps me to quickly understand if something is *really* related to a flow sheet or if the flow sheet is just one application of what is being addressed.

Commit:
More updates to the parameters.py module

- evaluation object property
- check bounds

Commit:
Initial evaluation object class (batch elution example)

Commit:
EvaluationPipeline (Batch Elution Example)

This is the version of the evaluation pipeline where the root argument
of the pipeline is found using ```pipeline.all_root_args``` from
pipefunc.

Commit:
Initial Evaluation Pipeline test suite
